lib/Widgets/HyperTextView.vala

```diff
@@ -0,0 +1,324 @@
+/*
+ *  Copyright (C) 2012–2021 elementary, Inc.
+ *
+ *  This program or library is free software; you can redistribute it
+ *  and/or modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 3 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General
+ *  Public License along with this library; if not, write to the
+ *  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301 USA.
+ */
+
+/**
+* This class enables navigatable URLs in Gtk.TextView
+*/
+public class Granite.HyperTextView : Gtk.TextView {
+
+    private const int FORCE_FULL_BUFFER_RESCAN_CHANGE_START_OFFSET = -1;
+
+    private uint buffer_changed_debounce_timeout_id = 0;
+    private int buffer_cursor_position_when_change_started = 0;
+
+    private GLib.HashTable<string, Gtk.TextTag> uri_text_tags;
+    private Regex uri_regex;
+
+    private bool is_control_key_pressed = false;
+
+    construct {
+        var http_charset = "[a-zA-Z0-9_\\/\\-\\+\\.:@\\?&%=#]";
+        var email_charset = "[a-zA-Z0-9_\\-\\.]";
+        var email_tld_charset = "[a-zA-Z0-9_\\-]";
+
+        var http_match_str = "https?:\\/\\/" + http_charset + "{1,}";
+        var email_match_str = "(mailto:)?" + email_charset + "{1,}@" + email_charset + "{1,}\\." + email_tld_charset + "{1,}";
```

Comment:
```suggestion
        var http_charset = "[\\w\\/\\-\\+\\.:@\\?&%=#]";
        var email_charset = "[\\w_\\-\\.]";
        var email_tld_charset = "[\\w_\\-]";

        var http_match_str = @"https?:\\/\\/$(http_charset)+?\\.$(http_charset)+";
        var email_match_str = @"(mailto:)?$(email_charset)+@$(email_charset)+\\.$(email_tld_charset)+";
```

You could simplify the regex a little bit. I also added a check that the URL has at least one `.`. If you don't want that, `http_match_string` should be
```vala
@"https?:\\/\\/$(http_charset)+"
```

Here's how it currently matches up against the list of URLs to test against:
![regex-negative](https://user-images.githubusercontent.com/29022469/133828899-b99b4f3c-68a4-4492-89d6-5f7ec4617ca0.png)
![regex-positive](https://user-images.githubusercontent.com/29022469/133828904-40344984-92c8-4f08-93b6-4711ee4c58db.png)